### PR TITLE
Fix generics creation time and allow model name reusing

### LINKED ISSUE #2069: Repeatedly subclassing a GenericModel with a Constrained Type fails
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())":`:
```
             pydantic version: 1.7.1
            pydantic compiled: False
                 install path: /home/al459/envs/pydantic/pydantic
               python version: 3.9.0 (default, Oct 27 2020, 13:45:54)  [GCC 9.3.0]
                     platform: Linux-5.4.0-52-generic-x86_64-with-glibc2.31
     optional deps. installed: []
```

I would like to create a generic model and subclass it with a constrained string type but I get a `TypeError: 'GenericBaseModel[ConstrainedStrValue]' already defined above, please consider reusing` if I test the script below:
```py
from typing import Generic, TypeVar

from pydantic import constr
from pydantic.generics import GenericModel

T = TypeVar("T")


class GenericBaseModel(GenericModel, Generic[T]):
    data: T


str20 = GenericBaseModel[constr(max_length=20)]
str10 = GenericBaseModel[constr(max_length=10)]
```
Error:
```py
――――――――――――――――――――――――――――――――――――――― ERROR collecting tests/test_custom.py ――――――――――――――――――――――――――――――――――――――――
NameError: Name conflict: 'GenericBaseModel[ConstrainedStrValue]' in 'tests.test_custom' is already used by <class 'te
sts.test_custom.GenericBaseModel[ConstrainedStrValue]'>

The above exception was the direct cause of the following exception:
tests/test_custom.py:14: in <module>
    ConstrB = GenericBaseModel[constr(max_length=10)]
pydantic/generics.py:82: in __class_getitem__
    raise TypeError(f'{model_name!r} already defined above, please consider reusing it') from NameError(
E   TypeError: 'GenericBaseModel[ConstrainedStrValue]' already defined above, please consider reusing it
```
The strange thing is if I wrap it in a function and test it the test passes:

```py
def test_generic_subclass_with_constrained_field():
    T = TypeVar("T")

    class GenericBaseModel(GenericModel, Generic[T]):
        data: T

    str20 = GenericBaseModel[constr(max_length=20)]
    str10 = GenericBaseModel[constr(max_length=10)]
```

### LINKED ISSUE #2070: [perfomance regression] Generic model creation cause huge slowdown
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

After upgrading pydantic to version 1.7 our slowdown detector founds a lot of slow python methods calls. After investigating stacktrace we found source. It is `get_caller_module_name()` (#1686) function which is extremely slow in _some_ enviroments. 

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.7
            pydantic compiled: True
                 install path: /tv/lib/python3.8/site-packages/pydantic
               python version: 3.8.6 (default, Oct  6 2020, 04:02:53)  [GCC 5.4.0 20160609]
                     platform: Linux-4.9.164-0409164-generic-x86_64-with-glibc2.17
     optional deps. installed: ['typing-extensions']
```
Note: HDD, docker enviroment with ubuntu14.04 LTS

```py
import typing

import pydantic.generics

CODE_T = typing.TypeVar('CODE_T')


class ResponseDetailCode(pydantic.generics.GenericModel, typing.Generic[CODE_T]):
	""
	detail: str
	code: CODE_T


with Timer() as t:  # just time.time wrapper
	for i in range(100):
		{'a': ResponseDetailCode[typing.Literal[f'foo{i}', f'bar{i}']]}
print(t.delta)
```

On 1.5.1 this code takes ~80ms. 
**On 1.7 - 1700ms (!).** 
So 1.7+ can not be used on production with HDD disks (on SSD it's like 20% slowdown).

Part of stack trace of class creation (from production env):
```
  File "/usr/lib/python3.8/posixpath.py", line 379, in abspath
    cwd = os.getcwd()
  File "/usr/lib/python3.8/inspect.py", line 721, in getabsfile
    return os.path.normcase(os.path.abspath(_filename))
  File "/usr/lib/python3.8/inspect.py", line 737, in getmodule
    file = getabsfile(object, _filename)
  File "/usr/lib/python3.8/inspect.py", line 708, in getsourcefile
    if getattr(getmodule(object, filename), '__loader__', None) is not None:
  File "/usr/lib/python3.8/inspect.py", line 1461, in getframeinfo
    filename = getsourcefile(frame) or getfile(frame)
  File "/usr/lib/python3.8/inspect.py", line 1491, in getouterframes
    frameinfo = (frame,) + getframeinfo(frame, context)
  File "/usr/lib/python3.8/inspect.py", line 1514, in stack
    return getouterframes(sys._getframe(1), context)
  File "/tv/lib/python3.8/site-packages/pydantic/generics.py", line 155, in get_caller_module_name
    previous_caller_frame = inspect.stack()[2].frame
  File "/tv/lib/python3.8/site-packages/pydantic/generics.py", line 65, in __class_getitem__
    model_module = get_caller_module_name() or cls.__module__
  File "***/views/models/permissions.py", line 43, in <module>
    ResponseDetailCode[Literal['...']]
```

It is obvious that  `inspect.stack()` reads many files/file stats from disk directly which is not normal behavior for "create class" operation.

### PR DESCRIPTION
**Pull request by MrMrRobat**
### Creation time:
- Use `sys._getframe()` instead of slow `inspect.stack()`
- Use `frame.f_globals` instead of slow `inspect.getmodule()` to retrieve caller module name 
- Combine `get_caller_module_name()` with  `is_call_from_module()` into more than 2000x 🙈 faster `get_caller_frame_info()` 

### Concrete names reusing:
- Append `'_'` to global reference name if it's already in use instead of raising `TypeError`

## Related issue number

fix #2070
fix #2069 

(updated: added "fix" so I don't forget to close both issues, @samuelcolvin)

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

File: pydantic/generics.py [hunk 1]
```diff
@@ -62,26 +60,26 @@ def __class_getitem__(cls: Type[GenericModelT], params: Union[Type[Any], Tuple[T
...
 
-        if is_call_from_module():  # create global reference and therefore allow pickling
-            object_in_module = sys.modules[model_module].__dict__.setdefault(model_name, created_model)
-            if object_in_module is not created_model:
-                # this should not ever happen because of _generic_types_cache, but just in case
-                raise TypeError(f'{model_name!r} already defined above, please consider reusing it') from NameError(
-                    f'Name conflict: {model_name!r} in {model_module!r} is already used by {object_in_module!r}'
-                )
+        if called_globally:  # create global reference and therefore allow pickling
+            object_by_reference = None
```
Comment (PrettyWood) [inline pydantic/generics.py:77]: Thanks for also taking care of this issue too besides performance one! 🎉

**Review by PrettyWood — COMMENTED** [review]
This probably needs to be added in v1.7.2 once last tests to keep coverage at 100% are added and @samuelcolvin approves

**Comment by MrMrRobat** [comment]
Coverage should be fixed now
Here's a benchmark showcasing the difference in speed between 1.7.1, 1.6.x and current PR, @bogdandm, could you test this one in your environment as well?
https://repl.it/repls/HorizontalConcernedWatchdog#main.py

**Comment by MrMrRobat** [comment]
Just decided to compare  speed of calling `get_caller_frame_info()` vs `get_caller_module_name()` with `is_call_from_module()`:
```py
import sys
from timeit import timeit

def get_caller_frame_info_new():
    previous_caller_frame = sys._getframe(2)
    frame_globals = previous_caller_frame.f_globals
    return frame_globals.get('__name__'), previous_caller_frame.f_locals is frame_globals

def get_caller_frame_info_old():
    # code of two functions placed in one and results are combined
    import inspect

    previous_caller_frame = inspect.stack()[2].frame
    previous_caller_module = inspect.getmodule(previous_caller_frame, previous_caller_frame.f_code.co_filename)
    name = previous_caller_module.__name__ if previous_caller_module is not None else None

    import inspect
    previous_caller_frame = inspect.stack()[2].frame
    called_globally = previous_caller_frame.f_locals is previous_caller_frame.f_globals

    return name, called_globally

def check():
    assert get_caller_frame_info_new() == get_caller_frame_info_old()

check()
old = timeit(get_caller_frame_info_old, number=1000)
new = timeit(get_caller_frame_info_new, number=1000)
print('new: ', new)
print('new: ', old)
print('old / new: ', old / new)
```

I'm getting this results
```py
new:  0.00040132100000001003
new:  1.176067545
old / new:  2930.490916249014
```
Three. Thousands. Times. Faster. ON SSD. 🙈 

I don't know what to say. Previous implementation is the definition of something unreasonably slow...

**Comment by samuelcolvin** [comment]
This is looking great, I'm going to play with it myself but it looks good.

Would also be good to wait for confirmation from @bogdandm.

**Comment by samuelcolvin** [comment]
Once this is merged, I'll release v1.7.2. If there's anything else we should include, let me know.

**Comment by bogdandm** [comment]
@MrMrRobat Could confirm that it is working (similar HDD setup, but a little bit faster disk I think)
```
new:  0.0005787028931081295
old:  1.042893337085843
old / new:  1802.1222107334108
```
And 2000 times faster on SSD as well.
Also just checked it by installing via pip from commit hash - it as fast as 1.5.1 with my code example from the issue.

**Comment by samuelcolvin** [comment]
Awesome. I'll make a new release in the morning.

**Pull request merged**